upf/upf_gtpu_encap.c

```diff
@@ -57,6 +57,238 @@ typedef enum
+  nlength = clib_host_to_net_u16 (length);
+  /* IE TEID I (5 bytes) + IE GTP Peer address (3 bytes + IPaddr length) + IE Recovery (2 bytes) */
+  b->current_length = 5 + 2 + 3 + length;
+
+  /* Encode TEID I, TEID already in NET order */
+  *p = 16; p++;
+  memcpy (p, &error->teid, sizeof (u32));
+  p += 4;
+  /* Encode Recovery */
+  *p = 14; p++;
+  *p = 0; p++;
+  /* Encode GTP Peer address */
+  *p = 133; p++;
+  memcpy (p, &nlength, sizeof (u16));
+  p += 2;
+  memcpy (p, (length == 4) ? (void *) &error->addr.ip4 : (void *) &error->addr.ip6, length);
+
+  return b->current_length; //bytes encoded
+}
+
+void
+upf_gtpu_error_ind (vlib_buffer_t * b0, int is_ip4)
+{
+  ip4_main_t *i4m = &ip4_main;
+  ip6_main_t *i6m = &ip6_main;
+  upf_main_t *gtm = &upf_main;
+  vlib_main_t *vm = gtm->vlib_main;
+  u32 bi = 0;
+  gtp_error_ind_t error;
+  vlib_buffer_t *p0;
+  u16 len_p1, new_len1, new_ip_len1;
+  ip_csum_t sum1;
+  u32 fib_index1;
+  u8 host_config_ttl = i4m->host_config.ttl;
+
+  if (vlib_buffer_alloc (vm, &bi, 1) != 1)
+    {
+      clib_warning ("No buffers available");
+      return;
+    }
+
+  p0 = vlib_get_buffer (vm, bi);
+  VLIB_BUFFER_TRACE_TRAJECTORY_INIT (p0);
+
+  /*
+   * b0 - received buffer, p0 - allocated buffer
+   * xxx0 header pointers used for b0
+   * xxx1 header pointers used for p1
+   */
+
+  if (is_ip4)
+    {
+      ip4_header_t *ip40, *ip41;
+      udp_header_t *udp0, *udp1;
+      gtpu_header_t *gtpu0, *gtpu1;
+
+      /* For ip46-lookup, set up FIB */
+      fib_index1 = vec_elt (i4m->fib_index_by_sw_if_index,
+			    vnet_buffer (b0)->sw_if_index[VLIB_RX]);
+      vnet_buffer (p0)->sw_if_index[VLIB_TX] = fib_index1;
+
+      /* Set buffer borders to include headers for now */
+      p0->current_length += sizeof (ip4_header_t) + sizeof (udp_header_t) + sizeof (gtpu_header_t);
+
+      vlib_buffer_reset (b0);
+      vlib_buffer_advance (b0, vnet_buffer (b0)->l4_hdr_offset);
+      gtpu0 = vlib_buffer_get_current (b0);
+
+      vlib_buffer_advance (b0, -(uword) (sizeof (ip4_header_t) + sizeof (udp_header_t)));
+      ip40 = vlib_buffer_get_current (b0);
+
+      udp0 = ip4_next_header (ip40);
+
+      ip41 = vlib_buffer_get_current (p0);
+      vlib_buffer_advance (p0, sizeof (ip4_header_t));
+      udp1 = vlib_buffer_get_current (p0);
+      vlib_buffer_advance (p0, sizeof (udp_header_t));
+      gtpu1 = vlib_buffer_get_current (p0);
+      // TODO: In a packet trace it looks fine, we can see p0 traced for TX
+      // But shouldn't we allocate new trace for it?
+      p0->trace_handle = b0->trace_handle;
+
+      /* Reuse IP settings of original packet */
+      memcpy (ip41, ip40, sizeof (ip4_header_t));
+
+      /* Swap addresses, save src addr to be encoded */
+      ip41->dst_address = ip40->src_address;
+      ip41->src_address = ip40->dst_address;
+      error.addr.ip4 = ip40->src_address;
+
+      /* Swap udp ports */
+      udp1->src_port = udp0->dst_port;
+      udp1->dst_port = udp0->src_port;
```

Comment:
from TS 29.281:
> The UDP destination port for the Error Indication shall be the user plane UDP port (2152).

Reply:
Encoded UDP port in a GTP Extension header UDP Port